### PR TITLE
[pythonic resources] Error early if a user annotates a param with a ConfigResourceFactory or bare ResourceDef

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Throws an error early if the user annotates an op, asset, schedule, or sensor function param with a `ConfigurableResourceFactory` or bare `ResourceDefinition` that is not a `ConfigurableResource`, e.g.

```python
class MyFactory(ConfigurableResourceFactory[str]):
    ...


# should be my_resource: Resource[str]!

@op
def my_op(my_resource: MyFactory):
    ...
```

## Test PLan

unit tests

### CODE REVIEW

**Comment by benpankow** [comment]
Current dependencies on/for this PR:
* master
  * **PR #13376** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13376" target="_blank">Graphite</a> 
    * **PR #13377** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13377" target="_blank">Graphite</a> 
      * **PR #13395** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/13395" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/13395?utm_source=stack-comment).

**Review by schrockn — APPROVED** [review]
cool

**Review by schrockn — CHANGES_REQUESTED** [review]
wait a sec there was an change submitted while i reviewed

**Review by schrockn — CHANGES_REQUESTED** [review]
So I generally think that boolean flags like this should be avoided. This leads future users of `get_resource_args` to ask: should I error aggressively? Why would I not check errors? Etc etc. In this case my interpretation is that you can only setting `err_aggressively=True` near entry points.

Instead I would recommend having a separate function (`validate_resource_annotatable_function` ?)  that does this type of error checking, and do it ASAP (e.g in the relevant `__init__` or immediately after the user passes the `fn` to the framework) and leave `get_resource_args` unmodified  Alternatively a separate function which does this more extensive error checking and then in turn calls `get_resource_args` that is named appropriately.

That you have to do all those dynamic imports is another "smell" that indicates that you might be jamming higher level logic into a lower layer piece of code.

**Review by schrockn — APPROVED** [review]

File: python_modules/dagster/dagster/_core/definitions/resource_annotation.py [hunk 1]
```diff
@@ -1,12 +1,52 @@
...
+                param_name=malformed_param.name,
+                annotation_type=malformed_param.annotation,
+                value_message=f"a '{output_type}'" if output_type else "an unknown",
+                annotation_suggestion=f"'Resource[{output_type.__name__}]'"
+                if output_type
+                else "'Resource[Any]' or 'Resource[<output type>]'",
+            )
+        )
+
+
```
Comment (schrockn) [inline python_modules/dagster/dagster/_core/definitions/resource_annotation.py:49]: consider moving this function to `structured_config` given the all the dynamic imports

**Review by schrockn — APPROVED** [review]

**Pull request merged**